base/strings/unicode.jl

```diff
@@ -669,6 +680,25 @@ function titlecase(s::AbstractString; wordsep::Function = !isletter, strict::Boo
     return String(take!(b))
 end
 
+# TODO: improve performance characteristics, room for a ~10x improvement.
+function titlecase(s::AnnotatedString; wordsep::Function = !isletter, strict::Bool=true)
+    state = (; startword = true, state = Ref{Int32}(0),
+             c0 = eltype(s)(zero(UInt32)), wordsep, strict)
+    annotated_chartransform(
+        s, function (c, state)
+            if isgraphemebreak!(state.state, state.c0, c) && state.wordsep(c)
+                state = Base.setindex(state, true, :startword)
+                cnew = c
+            else
+                cnew = state.startword ? titlecase(c) : state.strict ? lowercase(c) : c
+                state = Base.setindex(state, false, :startword)
+            end
+            state = Base.setindex(state, c, :c0)
+            cnew, state
+        end,
+        state)
```

Comment:
With my suggestion from above, perhaps could be rewritten like this (but perhaps I am missing something?)
```suggestion
    annotated_chartransform(s) do c
        if isgraphemebreak!(state.state, state.c0, c) && state.wordsep(c)
            state = Base.setindex(state, true, :startword)
            cnew = c
        else
            cnew = state.startword ? titlecase(c) : state.strict ? lowercase(c) : c
            state = Base.setindex(state, false, :startword)
        end
        state = Base.setindex(state, c, :c0)
        cnew
    end
```

Reply:
I've done this, but still explicitly passing `state`. I'm not sure if the explicit passing is really needed, but I'm worried about boxing.